subprojects/hashing/src/main/java/org/gradle/internal/hash/HashCode.java

```diff
@@ -95,98 +109,229 @@ private static int decode(char ch) {
+        }
+
+        @Override
+        public int hashCode() {
+            if (hashCode == 0) {
+                hashCode = (int) ((bytes[0] & 0xFF)
+                    | ((bytes[1] & 0xFF) << 8)
+                    | ((bytes[2] & 0xFF) << 16)
+                    | ((bytes[3] & 0xFF) << 24)
+                    // Make sure it's always > 0 but without affecting the lower 32 bits
+                    | (1L << 32));
+            }
+            return hashCode;
+        }
+
+        @Override
+        public boolean equals(@Nullable Object obj) {
+            if (obj == this) {
+                return true;
+            }
+
+            if (obj == null || obj.getClass() != ByteArrayBackedHashCode.class) {
+                return false;
+            }
+
+            byte[] a = bytes;
+            byte[] b = ((ByteArrayBackedHashCode) obj).bytes;
+            int length = a.length;
+
+            if (b.length != length) {
+                return false;
+            }
+
+            for (int i = 0; i < length; i++) {
+                if (a[i] != b[i]) {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        @Override
+        public int compareTo(@Nonnull HashCode o) {
+            return compareBytes(bytes, o.bytes());
+        }
+    }
+
+    // TODO Replace with Long.compare() after migrating off of Java 6
+    private static int compareLong(long a, long b) {
+        return (a < b) ? -1 : ((a == b) ? 0 : 1);
+    }
+
+    private static int compareBytes(byte[] a, byte[] b) {
         int result;
-        int len1 = bytes.length;
-        int len2 = bytes2.length;
+        int len1 = a.length;
+        int len2 = b.length;
         int length = Math.min(len1, len2);
         for (int idx = 0; idx < length; idx++) {
-            result = bytes[idx] - bytes2[idx];
+            result = a[idx] - b[idx];
             if (result != 0) {
                 return result;
             }
         }
         return len1 - len2;
     }
 
-    @Override
-    public String toString() {
-        StringBuilder sb = toStringBuilder(2 * bytes.length);
-        return sb.toString();
-    }
-
-    public String toZeroPaddedString(int length) {
-        StringBuilder sb = toStringBuilder(length);
-        while (sb.length() < length) {
-            sb.insert(0, '0');
-        }
-        return sb.toString();
-    }
-
-    private StringBuilder toStringBuilder(int capacity) {
-        StringBuilder sb = new StringBuilder(capacity);
-        for (byte b : bytes) {
-            sb.append(HEX_DIGITS[(b >> 4) & 0xf]).append(HEX_DIGITS[b & 0xf]);
-        }
-        return sb;
-    }
 
-    public String toCompactString() {
-        return new BigInteger(1, bytes).toString(36);
+    private static long bytesToLong(byte[] bytes, int offset) {
+        return (long) (bytes[offset] & 0xFF)
+            | (long) (bytes[offset + 1] & 0xFF) << 8
+            | (long) (bytes[offset + 2] & 0xFF) << 16
```

Comment:
Instead of casting to `long`, you can also use `& 0xFFL`.

Reply:
Thanks, I didn't know that. Now we do!